### PR TITLE
Add a latest Gauge type+collector support, import collector improvements

### PR DESCRIPTION
**Pull request by bickfordb**
* Update MmapedDict to optionally store timestamps
* Add a "latest" multi-process  Gauge: this persists the most recent value past worker process lifetime
* Update the merge/collector logic to support latest/min/max gauges
* Import the postal-main exporter code

### CODE REVIEW

File: prometheus_client/multiprocess.py [hunk 1]
```diff
@@ -122,7 +146,5 @@ def mark_process_dead(pid, path=None):
     """Do bookkeeping for when one process dies in a multi-process setup."""
     if path is None:
         path = os.environ.get('prometheus_multiproc_dir')
-    for f in glob.glob(os.path.join(path, 'gauge_livesum_{0}.db'.format(pid))):
-        os.remove(f)
-    for f in glob.glob(os.path.join(path, 'gauge_liveall_{0}.db'.format(pid))):
+    for f in glob.glob(os.path.join(path, 'gauge_{livelatest,liveall,livesum}_{0}.db'.format(pid))):
```
Comment (SysOfM) [inline prometheus_client/multiprocess.py]: This does not work as expected, I tried to kill the pid and get this error:

```
  File "/app/prometheus_client/multiprocess.py", line 149, in mark_process_dead
    for f in glob.glob(os.path.join(path, 'gauge_{livelatest,liveall,livesum}_{0}.db'.format(pid))):
KeyError: u'livelatest,liveall,livesum'
Logged from file glogging.py, line 271
```

File: prometheus_client/multiprocess.py [hunk 2]
```diff
@@ -56,7 +56,7 @@ def merge(files, accumulate=True):
                     metric.add_sample(name, labels_key, value)
             d.close()
 
-        for metric in metrics.values():
+        for n, metric in metrics.iteritems():
```
Comment (SysOfM) [inline prometheus_client/multiprocess.py]: `n` defined, but never used. I think it's worth returning `metrics.values ()`

Comment (bickfordb) [reply prometheus_client/multiprocess.py]: Good catch.  Fixed

File: tests/test_multiprocess.py [hunk 3]
```diff
@@ -73,31 +74,55 @@ def test_histogram_adds(self):
...
+        print os.listdir(os.environ["prometheus_multiproc_dir"])
+
         self.assertEqual(None, self.registry.get_sample_value('g', {'pid': '123'}))
         self.assertEqual(2, self.registry.get_sample_value('g', {'pid': '456'}))
 
+    def test_gauge_latest(self):
+        p0 = '123456'
+        p1 = '456789'
+        self.assertEqual(None, self.registry.get_sample_value('g'))
+        values.ValueClass = MultiProcessValue(lambda: p0)
```
Comment (SysOfM) [inline tests/test_multiprocess.py]: I suggest to make in the same style as the other tests.
`..lambda: '123')`
`('g', 'help'..`

```
values.ValueClass = MultiProcessValue(lambda: '123')
g1 = Gauge('g', 'help', registry=None, multiprocess_mode=Gauge.LATEST)
```

File: tests/test_multiprocess.py [hunk 4]
```diff
@@ -73,31 +74,55 @@ def test_histogram_adds(self):
...
+        g1 = Gauge('g', 'G', registry=None, multiprocess_mode=Gauge.LATEST)
+        t0 = time.time()
+        g1.set(1, timestamp=t0)
+        self.assertEqual(1, self.registry.get_sample_value('g'))
+        cleanup_dead_processes()
+        self.assertEqual(1, self.registry.get_sample_value('g'))
+        values.ValueClass = MultiProcessValue(lambda: p1)
+        g2 = Gauge('g', 'G', registry=None, multiprocess_mode=Gauge.LATEST)
```
Comment (SysOfM) [inline tests/test_multiprocess.py:119]: same as above

```
values.ValueClass = MultiProcessValue(lambda: '456')
g2 = Gauge('g', 'G', registry=None, multiprocess_mode=Gauge.LATEST)
```

File: tests/test_multiprocess.py [hunk 5]
```diff
@@ -73,31 +74,55 @@ def test_histogram_adds(self):
...
         values.ValueClass = MultiProcessValue(lambda: 456)
         g2 = Gauge('g', 'help', registry=None, multiprocess_mode='liveall')
-        self.assertEqual(0, self.registry.get_sample_value('g', {'pid': '123'}))
         self.assertEqual(0, self.registry.get_sample_value('g', {'pid': '456'}))
-        g1.set(1)
         g2.set(2)
         self.assertEqual(1, self.registry.get_sample_value('g', {'pid': '123'}))
         self.assertEqual(2, self.registry.get_sample_value('g', {'pid': '456'}))
         mark_process_dead(123, os.environ['prometheus_multiproc_dir'])
+        print os.listdir(os.environ["prometheus_multiproc_dir"])
```
Comment (SysOfM) [inline tests/test_multiprocess.py]: I think it is not necessary, debug information.
```suggestion
```

File: prometheus_client/multiprocess_exporter.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+
+def on_starting(server):
+    logging.basicConfig(stream=sys.stderr)
+    thread.start_new_thread(cleanup_thread, (), {})
+
+
+def app(req, start_response):
+    if req.get("PATH_INFO") == "/healthz":
+        body = "OK"
```
Comment (SysOfM) [inline prometheus_client/multiprocess_exporter.py]: I want to suggest returning a unique value on a healthz request

```suggestion
        body = json.dumps({'status': 'ok',
                           'timestamp': time.time()})
        
```

File: prometheus_client/multiprocess_exporter.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        headers = [("Content-Type", "text/plain"),
```
Comment (SysOfM) [inline prometheus_client/multiprocess_exporter.py]:
```suggestion
        headers = [("Content-Type", "application/json"),
```

File: prometheus_client/multiprocess_exporter.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
+import logging
+import sys
+import thread
+import time
+import traceback
+
```
Comment (SysOfM) [inline prometheus_client/multiprocess_exporter.py:5]: To use json in response body

```suggestion
import json

```

File: prometheus_client/multiprocess.py [hunk 9]
```diff
@@ -35,10 +45,10 @@ def merge(files, accumulate=True):
         """
         metrics = {}
         for f in files:
```
Comment (SysOfM) [inline prometheus_client/multiprocess.py:47]: I would like to suggest a slight simplification:

```
        for f in files:
            parts = os.path.splitext(os.path.basename(f))[0].split('_')
            typ = parts[0]
            multiprocess_mode = parts[1] if typ == 'gauge' else None
            pid = parts[2] if multiprocess_mode and len(parts) > 2 else None

            d = MmapedDict(f, read_mode=True)
            for key, value, timestamp in d.read_all_values():
                metric_name, name, labels = json.loads(key)
                labels_key = tuple(sorted(labels.items()))

                metric = metrics.get(metric_name)
                if metric is None:
                    metric = Metric(metric_name, 'Multiprocess metric', typ)
                    metrics[metric_name] = metric
                    
                if pid:
                    labels_key += (('pid', pid),)
                    metric._multiprocess_mode = multiprocess_mode
                    
                metric.add_sample(name, labels_key, value, timestamp=timestamp)
            d.close()
```

I do not insist, but I think it is easier to read :)

**Review by SysOfM — APPROVED** [review]
Looks good except few things.
Also please take a look at https://github.com/postmates/prometheus_client_python/pull/3

File: prometheus_client/multiprocess_exporter.py (hunk 6 above)
Comment (bickfordb) [reply prometheus_client/multiprocess_exporter.py]: Like a timestamp?

File: prometheus_client/multiprocess_exporter.py (hunk 7 above)
Comment (bickfordb) [reply prometheus_client/multiprocess_exporter.py]: I think this is fine without a JSON object as the response?  This is just for a health check looking for a successful HTTP response (200 status code)

File: tests/test_multiprocess.py (hunk 3 above)
Comment (bickfordb) [reply tests/test_multiprocess.py]: Ok, fixed this

File: prometheus_client/multiprocess_exporter.py (hunk 6 above)
Comment (SysOfM) [reply prometheus_client/multiprocess_exporter.py]: yep

File: prometheus_client/multiprocess_exporter.py (hunk 7 above)
Comment (SysOfM) [reply prometheus_client/multiprocess_exporter.py]: Yes, of course, but it is better to get a unique value to be sure that the service is working.  But I do not insist on this, at your discretion.

Comment (bickfordb) [reply prometheus_client/multiprocess_exporter.py]: This actually isn't being checked any longer by the Kubernetes probe, so I'll just remove this for simplicity

**Pull request merged**